### PR TITLE
Upgrade enhancements

### PR DESCRIPTION
**Pull request by niladrih**
Changes:
- Refactor crate::common::kube_client.rs to use new module structure crate::common::kube:client.rs
- Extract pagination logic for k8s_openapi/kube-rs listing to a different function
- Add HELM_DRIVER env to helm commands
- Pass HELM_DRIVER, RUST_LOG env values from kubectl-mayastor to upgrade-job
- Detect dependency mayastor's version from openebs helm chart via helm storage driver read
- Support secret and configmap helm storage driver
- Add support for helm's `--reset-then-reuse-values` flag via `kubectl mayastor upgrade --reset-then-reuse-values`
- Add Gzip decompression tools and base64 decoding tools to decode helm storage driver data.
- Add RBAC for listing, getting controllerrevisions on the kubectl plugin for the upgrade-job
- Modify openebs chart detection regex to include helm_core style `-0-main-unstable-timestamp-0` suffix on the semver.

### CODE REVIEW

File: chart/shell.nix [hunk 1]
```diff
@@ -14,7 +14,7 @@ mkShell {
     coreutils
     git
     helm-docs
-    kubernetes-helm-wrapped
+    kubernetes-helm
```
Comment (tiagolobocastro) [inline chart/shell.nix]: why this change?

File: k8s/upgrade/Cargo.toml [hunk 2]
```diff
@@ -39,3 +39,5 @@ tokio = { version = "1.33.0", features = ["full"] }
 tempfile = "3.8.0"
 # Tracing
 tracing = "0.1.37"
+base64 = "0.22.1"
+flate2 = "1.0.33"
```
Comment (tiagolobocastro) [inline k8s/upgrade/Cargo.toml]: Use existing versions

File: k8s/upgrade/src/bin/upgrade-job/common/kube/client.rs [hunk 3]
```diff
@@ -217,3 +229,137 @@ pub(crate) async fn latest_controller_revision_hash(
...
+    }
+
+    let list_secrets_error = ListSecretsWithLabelAndField {
+        label: label_selector.unwrap_or_default(),
+        field: field_selector.unwrap_or_default(),
+        namespace: namespace.clone(),
+    };
+
+    loop {
+        let secrets_list = secrets_api(namespace.as_str())
```
Comment (tiagolobocastro) [inline k8s/upgrade/src/bin/upgrade-job/common/kube/client.rs]: get secrets api only once and reuse?

File: k8s/upgrade/src/bin/upgrade-job/common/kube/client.rs [hunk 4]
```diff
@@ -217,3 +229,137 @@ pub(crate) async fn latest_controller_revision_hash(
...
+}
+
+/// GET the helm release secret for a helm release in a namespace.
+pub(crate) async fn get_helm_release_secret(
+    release_name: String,
+    namespace: String,
+) -> Result<Secret> {
+    let secrets = list_secrets(
+        namespace.clone(),
+        Some(format!("name={}", release_name.as_str())),
```
Comment (tiagolobocastro) [inline k8s/upgrade/src/bin/upgrade-job/common/kube/client.rs]:
```suggestion
        Some(format!("name={release_name}")),
```

File: k8s/upgrade/src/bin/upgrade-job/common/kube/client.rs [hunk 5]
```diff
@@ -217,3 +229,137 @@ pub(crate) async fn latest_controller_revision_hash(
...
+}
+
+/// GET the helm release configmap for a helm release in a namespace.
+pub(crate) async fn get_helm_release_configmap(
+    release_name: String,
+    namespace: String,
+) -> Result<ConfigMap> {
+    let cms = list_configmaps(
+        namespace.clone(),
+        Some(format!("name={},owner=helm", release_name.as_str())),
```
Comment (tiagolobocastro) [inline k8s/upgrade/src/bin/upgrade-job/common/kube/client.rs]:
```suggestion
        Some(format!("name={release_name},owner=helm")),
```

File: k8s/upgrade/src/bin/upgrade-job/helm/client.rs [hunk 6]
```diff
@@ -1,24 +1,108 @@
...
+            .data
+            .ok_or(HelmStorageNoData { driver }.build())?
+            .into_iter()
+            .find_map(|(k, v)| k.eq("release").then_some(v))
+            .ok_or(HelmStorageNoReleaseValue { driver }.build())
+    }};
+}
+
+#[derive(Debug, Deserialize)]
+pub(crate) struct HelmChartRelease {
```
Comment (tiagolobocastro) [inline k8s/upgrade/src/bin/upgrade-job/helm/client.rs:37]: missing doc comments throughout

File: k8s/upgrade/src/bin/upgrade-job/helm/client.rs [hunk 7]
```diff
@@ -60,6 +156,7 @@ impl HelmReleaseClientBuilder {
 #[derive(Clone)]
 pub(crate) struct HelmReleaseClient {
     pub(crate) namespace: String,
+    pub(crate) storage_driver: String,
```
Comment (tiagolobocastro) [inline k8s/upgrade/src/bin/upgrade-job/helm/client.rs:168]: what is this storage_driver? add doc comment

File: k8s/upgrade/src/bin/upgrade-job/helm/client.rs [hunk 8]
```diff
@@ -169,6 +268,46 @@ impl HelmReleaseClient {
...
+                debug!("Using helm configmap as helm storage");
+                let cm = KubeClient::get_helm_release_configmap(
+                    release_name.to_string(),
+                    self.namespace.clone(),
+                )
+                .await?;
+                let release_data = extract_data!(cm)?;
+                let decoded_data = decode_decompress_data(release_data)?;
+                let dependencies = dependencies_from_release_data(decoded_data)?;
+                debug!(data=?dependencies, "Found helm chart release chart metadata dependency in helm configmap");
```
Comment (tiagolobocastro) [inline k8s/upgrade/src/bin/upgrade-job/helm/client.rs:312]: Can you control the RUST_LOG for the upgrade job via kubectl-mayastor upgrade?

File: k8s/upgrade/src/bin/upgrade-job/helm/upgrade.rs [hunk 9]
```diff
@@ -178,6 +193,15 @@ impl HelmUpgraderBuilder {
 
         // Determine chart variant.
         if Regex::new(umbrella_regex.as_str())?.is_match(chart.as_ref()) {
+            // The version of the Core helm chart (installed as a dependent chart) which is
+            // installed in the cluster.
+            let source_version =
+                core_version_from_umbrella_release(&client, helm_release.name()).await?;
+            debug!(version=%source_version, "Found version of dependent chart {CORE_CHART_NAME}");
```
Comment (tiagolobocastro) [inline k8s/upgrade/src/bin/upgrade-job/helm/upgrade.rs]: These would be useful as info?

```suggestion
            info!(version=%source_version, "Found version of dependent chart {CORE_CHART_NAME}");
```

File: k8s/upgrade/src/bin/upgrade-job/helm/upgrade.rs [hunk 10]
```diff
@@ -188,12 +212,14 @@ impl HelmUpgraderBuilder {
                 target_version,
             }))
         } else if Regex::new(core_regex.as_str())?.is_match(chart) {
+            // The version of the Core helm chart (installed as the parent chart)
+            // which is installed in the cluster.
+            let source_version = version_from_core_chart_release(chart)?;
```
Comment (tiagolobocastro) [inline k8s/upgrade/src/bin/upgrade-job/helm/upgrade.rs:223]: Also add info log here?

File: k8s/upgrade/src/bin/upgrade-job/opts.rs [hunk 11]
```diff
@@ -41,21 +41,29 @@ pub(crate) struct CliArgs {
...
     fmt_style: FmtStyle,
 
     /// Use ANSI colors for the logs.
-    #[clap(long)]
+    #[arg(long, default_value_t = true)]
     ansi_colors: bool,
+
+    /// This is the helm storage driver, e.g. secret, configmap, memory, etc.
+    #[arg(env = "HELM_DRIVER", default_value = "")]
+    helm_storage_driver: String,
```
Comment (tiagolobocastro) [inline k8s/upgrade/src/bin/upgrade-job/opts.rs:62]:
```suggestion
    #[arg(env = "HELM_DRIVER)]
    helm_storage_driver: Option<String>,
```

File: chart/shell.nix (hunk 1 above)
Comment (niladrih) [reply chart/shell.nix]: It seems like the package `kubernetes-helm-wrapped` doesn't exist in NixOS/nixpkgs anymore. This code change in the upgrade-job requires helm 3.14 or above. This change dependends on the nixpkgs upgrade from 22.11 to 24.05 (helm 3.15.0). The package name for the helm nixpkg is `kubernetes-helm`.

Comment (niladrih) [reply chart/shell.nix]: The upgrade to nixpkgs 24.05 seems to not be a straighforward plug and play. I'll need some help with that.

File: k8s/upgrade/src/bin/upgrade-job/common/kube/client.rs (hunk 3 above)
Comment (niladrih) [reply k8s/upgrade/src/bin/upgrade-job/common/kube/client.rs]: Sure..

File: k8s/upgrade/src/bin/upgrade-job/helm/client.rs (hunk 6 above)
Comment (niladrih) [reply k8s/upgrade/src/bin/upgrade-job/helm/client.rs:37]: Yess! I just finished testing this code. I'll add those in!

File: k8s/upgrade/src/bin/upgrade-job/helm/client.rs (hunk 8 above)
Comment (niladrih) [reply k8s/upgrade/src/bin/upgrade-job/helm/client.rs:312]: That's a good point, I'd guess no. That's a worthwhile change. I'll add it in.

File: k8s/upgrade/src/bin/upgrade-job/opts.rs (hunk 11 above)
Comment (niladrih) [reply k8s/upgrade/src/bin/upgrade-job/opts.rs:62]: This is done to keep the logic close to how the helm codebase does it. I'd prefer to keep this how it is currently and add comments to the relevant lines.

Ref:
- https://github.com/helm/helm/blob/v3.15.0/cmd/helm/helm.go#L74
- https://pkg.go.dev/os#Getenv

WDYT?

File: chart/shell.nix (hunk 1 above)
Comment (tiagolobocastro) [reply chart/shell.nix]: Sure, let's sync up offline

File: k8s/upgrade/src/bin/upgrade-job/helm/client.rs (hunk 8 above)
Comment (niladrih) [reply k8s/upgrade/src/bin/upgrade-job/helm/client.rs:312]: I've added a change to read the env from the shell, and not picking it up from the plugin's tracing config.

**Comment by niladrih** [comment]
Reverted the kubernetes-helm-wrapped change.

File: k8s/upgrade/src/bin/upgrade-job/opts.rs [hunk 12]
```diff
@@ -41,21 +41,29 @@ pub(crate) struct CliArgs {
 
     /// The set values specified by the user for upgrade
     /// (can specify multiple or separate values with commas: key1=val1,key2=val2).
-    #[arg(long)]
+    #[arg(long, default_value = "")]
```
Comment (tiagolobocastro) [inline k8s/upgrade/src/bin/upgrade-job/opts.rs:44]: Should these be Option<String>

File: k8s/upgrade/src/plugin/objects.rs [hunk 13]
```diff
@@ -338,6 +348,12 @@ pub(crate) fn upgrade_job(
                                 }),
                                 ..Default::default()
                             },
+                            EnvVar {
+                                // Ref: https://github.com/helm/helm/blob/main/cmd/helm/helm.go#L76
+                                name: "HELM_DRIVER".to_string(),
+                                value: Some(env::var("HELM_DRIVER").unwrap_or_default()),
```
Comment (tiagolobocastro) [inline k8s/upgrade/src/plugin/objects.rs]: Now I'm confused, you made point about the cli having default value for the helm storage driver, but here you're ignoring the cli argument entirely and just using the env variable

File: tests/bdd/common/helm.py [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+
+    try:
+        repo_name = "mayastor"
+        subprocess.run(
+            [
+                "helm",
+                "repo",
+                "add",
+                repo_name,
+                "https://openebs.github.io/mayastor-extensions",
```
Comment (tiagolobocastro) [inline tests/bdd/common/helm.py]: Shouldn't we install the local chart?

File: tests/bdd/common/helm.py [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+    def install_mayastor(self, source: ChartSource, version=None):
+        if self.release_is_deployed("mayastor"):
+            print(
+                f"WARN: Helm release 'mayastor' already exists in the 'mayastor' namespace."
+            )
+            return
+
+        install_command = []
+        if source == ChartSource.HOSTED:
+            repo_name = repo_add_mayastor()
```
Comment (tiagolobocastro) [inline tests/bdd/common/helm.py]: ah ok, we want to be able to do both, got it.
Let's change the script to support this as well, rather than have separate commands here

File: tests/bdd/features/upgrade.feature [hunk 16]
```diff
@@ -0,0 +1,8 @@
+Feature: Upgrade
+
+  Background:
+    Given an installed mayastor helm chart
+
+  Scenario: upgrade command is issued
+    When a kubectl mayastor upgrade command is issued to
```
Comment (tiagolobocastro) [inline tests/bdd/features/upgrade.feature]:
```suggestion
    When a kubectl mayastor upgrade command is issued
```

**Review by tiagolobocastro — COMMENTED** [review]
Add some more details to your PR's description please.
Looks like commit lint needs to be fixed and tests also failing

File: k8s/upgrade/src/plugin/objects.rs (hunk 13 above)
Comment (niladrih) [reply k8s/upgrade/src/plugin/objects.rs]: Aaaaah, good catch!

Comment (niladrih) [reply k8s/upgrade/src/plugin/objects.rs]: Oh, I'm sorry, that's not what's happening here. The CLI arg is an env for the upgrade job. The env is passed down from the kubectl-mayastor plugin to the upgrade-job.

Comment (tiagolobocastro) [reply k8s/upgrade/src/plugin/objects.rs]: Oh, I see!
```suggestion
                                value: env::var("HELM_DRIVER").ok(),
```

File: k8s/upgrade/src/bin/upgrade-job/opts.rs (hunk 12 above)
Comment (niladrih) [reply k8s/upgrade/src/bin/upgrade-job/opts.rs:44]: This is used as a suffix to a command &str/String literal. The empty String makes it easy to append it at the end of the upgrade command. Option would require unwrap blocks, etc. 

So no real reason for this, other than convenience. I'd like to leave this in for now. I am adding this in in this PR so that the binary is usable without these options.

File: tests/bdd/common/helm.py (hunk 15 above)
Comment (niladrih) [reply tests/bdd/common/helm.py]: Do we need that, it's a well known helm command. Maybe we leave this in for the time being, and export it to the install.sh when another use-case comes along?

WDYT?

Comment (niladrih) [reply tests/bdd/common/helm.py]: I'll put this comment in the tests PR 564.

**Review by tiagolobocastro — APPROVED** [review]

**Review by Abhinandan-Purkait — APPROVED** [review]

**Comment by niladrih** [comment]
bors merge
